OutOfSchool/OutOfSchool.WebApi.Tests/Services/NotificationServiceTests.cs

```diff
@@ -0,0 +1,270 @@
+    }
+
+    [Test]
+    public void ReadAll_WhenUserIsValidAndDbHasConcurrencyViolation_ShouldThrowDbUpdateConcurrencyException()
+    {
+        // Arrange
+        notificationRepositoryMock
+            .Setup(n => n.SetReadDateTimeForAllUnreaded(It.IsAny<string>(), It.IsAny<DateTimeOffset>()))
+            .Throws<DbUpdateConcurrencyException>();
+
+        // Act & Assert
+        Assert.ThrowsAsync<DbUpdateConcurrencyException>(() => notificationService.ReadAll(userId));
+    }
+    #endregion
+
+    #region GetAllUsersNotificationsGroupedAsync
+    [Test]
+    public async Task GetAllUsersNotificationsGroupedAsync_WithValidUserId_ShouldReturnGroupedNotifications()
+    {
+        // Arrange
+        var notificationsConfig = new NotificationsConfig
+        {
+            Grouped = new List<string> { "Application", },
+        };
+
+        notificationsConfigMock.Setup(x => x.Value).Returns(notificationsConfig);
+
+        notificationRepositoryMock
+            .Setup(x => x.GetByFilter(
+                It.IsAny<Expression<Func<Notification, bool>>>(),
+                It.IsAny<string>()))
+            .Returns(Task.FromResult<IEnumerable<Notification>>(notifications));
+
+        var expectedGrouped = notifications
+            .Where(n => notificationsConfig.Grouped.Contains(n.Type.ToString()))
+            .GroupBy(n => new { n.Type, n.Action, n.GroupedData })
+            .Select(n => new NotificationGrouped
+            {
+                Type = n.Key.Type,
+                Action = n.Key.Action,
+                GroupedData = n.Key.GroupedData,
+                Amount = n.Count(),
+            })
+            .ToList();
+
+        var expectedGroupedByType = expectedGrouped
+            .GroupBy(n => n.Type)
+            .Select(n => new NotificationGroupedByType
+            {
+                Type = n.Key,
+                Amount = n.Sum(g => g.Amount),
+                GroupedByAdditionalData = n.ToList(),
+            })
+            .ToList();
+
+        var expectedNotifications = notifications
+            .Where(n => !notificationsConfig.Grouped.Contains(n.Type.ToString()))
+            .Select(notification => mapper.Map<NotificationDto>(notification))
+            .OrderByDescending(n => n.CreatedDateTime)
+            .ToList();
+
+        // Act
+        var result = await notificationService.GetAllUsersNotificationsGroupedAsync(userId).ConfigureAwait(false);
+
+        // Assert
+        notificationRepositoryMock.VerifyAll();
+        TestHelper.AssertTwoCollectionsEqualByValues(expectedNotifications, result.Notifications);
+        TestHelper.AssertTwoCollectionsEqualByValues(
+            expectedGroupedByType.Select(x => x.Amount),
+            result.NotificationsGroupedByType.Select(x => x.Amount));
+        Assert.AreEqual(
+            expectedGroupedByType.Select(x => x.GroupedByAdditionalData).Count(),
+            result.NotificationsGroupedByType.Select(x => x.GroupedByAdditionalData).Count());
```

Comment:
(Excessive code) `collection.Select(...).Count()` could be simplified to `collection.Count()`

Reply:
> (Excessive code) `collection.Select(...).Count()` could be simplified to `collection.Count()`

Fixed. Thank you!